### PR TITLE
Windows: Don't use Optlink by default

### PR DESCRIPTION
**Pull request by wilzbach**
There's __a lot__ of evidence that the Windows experience of D is frustrating as almost every big Dub package doesn't compile with Optlink and the user manually needs to figure out to switch to COFF.

This PR proposes to use COFF by default on Windows.

Evidence collection:
--------------------------

- https://github.com/vibe-d/vibe.d/issues/2243
- https://forum.dlang.org/thread/bobqtoesasjlwvdiisih@forum.dlang.org
- https://forum.dlang.org/post/wpxsfavlgqtfqplwizme@forum.dlang.org
- https://forum.dlang.org/post/hoyaqzytksiuvdlsitaw@forum.dlang.org
- https://forum.dlang.org/post/uaprrobsefqhaqahcthu@forum.dlang.org
- https://forum.dlang.org/post/xbhsbhepqcnfbcveuduw@forum.dlang.org
- https://forum.dlang.org/post/xbzqqhzjrjxndzgqogyk@forum.dlang.org
- https://issues.dlang.org/show_bug.cgi?id=17508
- https://github.com/vibe-d/vibe.d/issues/2122
- https://github.com/vibe-d/vibe.d/issues/1910
- https://github.com/vibe-d/vibe.d/issues/1771
- https://github.com/vibe-d/vibe.d/issues/1439
- https://forum.dlang.org/post/owoazhliszwqcqvsnifl@forum.dlang.org
- https://forum.dlang.org/post/abxmmjwpwmelxyicuzip@forum.dlang.org
- https://forum.dlang.org/post/zjhvufwmratwxcpywjsg@forum.dlang.org
- https://forum.dlang.org/post/ybufckdvasbnehwbdnec@forum.dlang.org
- https://forum.dlang.org/post/gemkzmxsdyilziloorqv@forum.dlang.org
- https://forum.dlang.org/post/yawsuwxbzkdlboniyyio@forum.dlang.org
- https://forum.dlang.org/post/ryzmhycnzowbyurvnlsu@forum.dlang.org
...

TODO:
- [x] changelog entry
- [x] Test on Windows

I don't have a Windows box at the moment, so if one of the many devs who actually run into this issue would step to test this, this would be awesome.

### CODE REVIEW

**Comment by andre2007** [comment]
source\dub\compilers\dmd.d(48): Error: incompatible types for (isWow64Process) !is (0): extern (Windows) int function() nothrow and int
source\dub\compilers\dmd.d(51): Error: function pointer isWow64Process() is not callable using argument types (void*, bool*)
source\dub\compilers\dmd.d(51):        expected 0 argument(s), not 2
source\dub\compilers\dmd.d(123): Error: undefined identifier isWow64, did you mean function IsWow64?

File: source/dub/compilers/dmd.d [hunk 1]
```diff
@@ -22,6 +22,37 @@ import std.file;
 import std.process;
 import std.typecons;
 
+version (Windows)
```
Comment (andre2007) [inline source/dub/compilers/dmd.d]: It works if you replace this coding with this one:

```d
// Determines whether the specified process is running under WOW64 or an Intel64 of x64 processor.
version (Windows)
private Nullable!bool isWow64() {
	// See also: https://docs.microsoft.com/en-us/windows/desktop/api/wow64apiset/nf-wow64apiset-iswow64process
	import core.sys.windows.basetsd : HANDLE;
	import core.sys.windows.winbase : GetProcAddress, GetModuleHandleA, GetCurrentProcess;
	
	alias isWow64Process = extern (Windows) nothrow @nogc bool function(HANDLE handle, bool* isWow64);
	static Nullable!bool result;

	// A process's architecture won't change over while the process is in memory
	// Return the cached result
	if (!result.isNull)
		return result;

	auto isWow64ProcessFp = cast(isWow64Process) GetProcAddress(GetModuleHandleA("kernel32"), "IsWow64Process");

	if(isWow64ProcessFp !is null)
	{
		bool bIsWow64;
		if (isWow64ProcessFp(GetCurrentProcess(), &bIsWow64))
			result = bIsWow64;
	}
	return result;
}
```

**Comment by andre2007** [comment]
WIth the proposed changes, it works fine on windows. Overwriting the defaulted architecture with command line arg "arch" also works fine.

**Comment by andre2007** [comment]
@wilzbach I tried to create a pull request against your fork of dub to propose the mentioned changes, but github instead changed the source code directly. That was not my intention, sorry!

**Comment by wilzbach** [comment]
No, it's totally fine to push to my PRs. It saves everyone time. Thanks a lot for helping out in this issue!

**Comment by wilzbach** [comment]
Added a changelog entry, so this should be ready to go.
CC @thewilsonator @s-ludwig

**Comment by s-ludwig** [comment]
Shouldn't this use [`GetSystemInfo`](https://docs.microsoft.com/en-us/windows/desktop/api/sysinfoapi/nf-sysinfoapi-getsysteminfo) instead (the `SYSTEM_INFO.wProcessorArchitecture` field)? The current code will only work correctly if DUB is compiled as a 32-bit executable, AFAICS.

`GetSystemInfo` does not require `LoadLibrary`, which would also shorten the code quite a bit.

**Comment by andre2007** [comment]
As proposed by @s-ludwig 

```d
// Determines whether the specified process is running under WOW64 or an Intel64 of x64 processor.
version (Windows)
private Nullable!bool isWow64() {
	// See also: https://docs.microsoft.com/de-de/windows/desktop/api/sysinfoapi/nf-sysinfoapi-getnativesysteminfo
	import core.sys.windows.windows : GetNativeSystemInfo, SYSTEM_INFO, PROCESSOR_ARCHITECTURE_AMD64;
	
	static Nullable!bool result;
	
	// A process's architecture won't change over while the process is in memory
	// Return the cached result
	if (!result.isNull)
		return result;
	
	SYSTEM_INFO systemInfo;
	GetNativeSystemInfo(&systemInfo);
	result = systemInfo.wProcessorArchitecture == PROCESSOR_ARCHITECTURE_AMD64;
	return result;
}
```

I also tried ```GetSystemInfo``` but that returned 0 (PROCESSOR_ARCHITECTURE_INTEL) on my system.

File: source/dub/compilers/dmd.d [hunk 2]
```diff
@@ -86,7 +111,14 @@ config    /etc/dmd.conf
 		string[] arch_flags;
 		switch (arch_override) {
 			default: throw new Exception("Unsupported architecture: "~arch_override);
-			case "": break;
+			case "":
+				// Don't use Optlink by default on Windows
+				version (Windows) {
+					const is64bit = isWow64();
+					if (!is64bit.isNull)
```
Comment (WebFreak001) [inline source/dub/compilers/dmd.d]: shouldn't this also check the value instead of just if it's null? `isWow64ProcessFp` could also return false.

**Comment by WebFreak001** [comment]
wouldn't a `version (Win64)` with `return true` be enough?

**Comment by wilzbach** [comment]
> wouldn't a `version (Win64)` with `return true` be enough?

No, because the `version` is taken from the compiler which compiles dub and currently we don't even ship a `DMD64` compiler, so I assume neither do we ship a dub 64-bit executable. So this would still result in 32-bit still being the default on 64-bit machines.

**Comment by wilzbach** [comment]
@andre2007 thanks a lot! I updated the PR, but as mentioned I haven't used a Windows machine for years and thus I don't have a development machine setup nor a huge interest compared to people actually using Windows in this, so please feel free to push to this PR. I solely opened it as a PoC to show how easy it would be to change the defaults, because I was annoyed at replying to people to use `-m64` due to OPTLINK build failures.

**Comment by andre2007** [comment]
Could we merge this pr?

**Comment by thewilsonator** [comment]
This good to go?

**Comment by andre2007** [comment]
@wilzbach with latest commit a syntax error was introduced.
systemInfo is now undefined

**Comment by thewilsonator** [comment]
I fixed that, @wilzbach can we get auto-merge-squash for all the dlang repos (or someone can squash this when its green).

**Comment by wilzbach** [comment]
> I fixed that

Thanks!

> @wilzbach can we get auto-merge-squash for all the dlang repos (or someone can squash this when its green).

Sure. Added.

**Comment by thewilsonator** [comment]
Thanks!

**Comment by thewilsonator** [comment]
Hmm any idea why 32-bit deployment failed? https://travis-ci.org/dlang/dub/jobs/509399958

**Comment by wilzbach** [comment]
This 32-bit deployment is also broken on master: https://travis-ci.org/dlang/dub

**Comment by kinke** [comment]
So if I run a 32-bit Windows on a non-ancient CPU, dub now defaults to a 64-bit binary which doesn't run on the dev system?

**Pull request merged**